### ISSUE TITLE
deepcopy broken for ItemSortedDict

### ISSUE THREAD

**Issue by basnijholt** [grantjenks/python-sortedcollections#5]
Try the following snippet:
```python
import sortedcollections
import math
from copy import deepcopy

def finite_loss(ival, loss, xscale):
    if math.isinf(loss):
        loss = (ival[1] - ival[0]) / xscale
    return loss, ival

def loss_manager(xscale):
    def sort_key(ival, loss):
        loss, ival = finite_loss(ival, loss, xscale)
        return -loss, ival
    return sortedcollections.ItemSortedDict(sort_key)

lm = loss_manager(xscale=1)
lm2 = deepcopy(lm)
lm2[0, 0] = 1
```

This has the traceback
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-2d05e49f7b81> in <module>
     24 lm = loss_manager(1)
     25 lm2 = deepcopy(lm)
---> 26 lm2[0, 0] =1

~/miniconda3/lib/python3.6/site-packages/sortedcollections/recipes.py in __setitem__(self, key, value)
     81             self._dict_delitem(key)
     82         self._dict_setitem(key, value)
---> 83         self._list_add(key)
     84 
     85     def copy(self):

~/miniconda3/lib/python3.6/site-packages/sortedcontainers/sortedlist.py in add(self, value)
   1781         _maxes = self._maxes
   1782 
-> 1783         key = self._key(value)
   1784 
   1785         if _maxes:

~/miniconda3/lib/python3.6/site-packages/sortedcollections/recipes.py in key_func(key)
     64         def key_func(key):
     65             "Apply key function to (key, value) item pair."
---> 66             return func(key, self[key])
     67         args[0] = key_func
     68         super(ItemSortedDict, self).__init__(*args, **kwargs)

TypeError: key_func() takes 1 positional argument but 2 were given
```

**Comment by jbweston** [comment]
An even more condensed example that produces the same error:

```python
from copy import deepcopy
import sortedcollections as sc

d = sc.ItemSortedDict(lambda *args: args)

# get/set item works
d['hello'] = 'world'
d['hello']

deepcopy(d)  # raises error
```

Observer with `sortedcollections` 1.0.1

**Comment by basnijholt** [comment]
I've implemented [a temporary solution in `adaptive`](https://gitlab.kwant-project.org/qt/adaptive/merge_requests/139/diffs?commit_id=b3d315bf1d6cb0cf64691a613a65d524f7d509fb) where I monkey patch the `ItemSortedDict`:

```python
def fix_deepcopy(sorted_dict, xscale):
    # XXX: until https://github.com/grantjenks/sortedcollections/issues/5 is fixed
    import types
    from copy import deepcopy
    def __deepcopy__(self, memo):
        items = deepcopy(list(self.items()))
        lm = loss_manager(self.xscale)
        lm.update(items)
        return lm
    sorted_dict.xscale = xscale
    sorted_dict.__deepcopy__ = types.MethodType(__deepcopy__, sorted_dict)

def loss_manager(xscale):
    def sort_key(ival, loss):
        loss, ival = finite_loss(ival, loss, xscale)
        return -loss, ival
    sorted_dict = sortedcollections.ItemSortedDict(sort_key)
    fix_deepcopy(sorted_dict, xscale)
    return sorted_dict
```

**Comment by grantjenks** [comment]
Fixed and deployed at v1.1.1
